packages/labs/compiler/src/lib/ast-fragments.ts

```diff
@@ -7,14 +7,155 @@
+  ATTRIBUTE: 1,
   CHILD: 2,
+  PROPERTY: 3,
+  BOOLEAN_ATTRIBUTE: 4,
+  EVENT: 5,
+  ELEMENT: 6,
+  COMMENT_PART: 7,
 } as const;
 
 export type PartType = (typeof PartType)[keyof typeof PartType];
 
-export type TemplatePart = {
-  type: typeof PartType.CHILD;
-  index: number;
+export type TemplatePart =
+  | {
+      type:
+        | typeof PartType.CHILD
+        | typeof PartType.COMMENT_PART
+        | typeof PartType.ELEMENT;
+      index: number;
+    }
+  | {
+      type: typeof PartType.ATTRIBUTE;
+      index: number;
+      name: string;
+      strings: Array<string>;
+      ctorType: PartType;
+    };
+
+const attributePartConstructors = {
+  [PartType.ATTRIBUTE]: 'AttributePart',
+  [PartType.PROPERTY]: 'PropertyPart',
+  [PartType.BOOLEAN_ATTRIBUTE]: 'BooleanAttributePart',
+  [PartType.EVENT]: 'EventPart',
+} as const;
+
+// These constructors have been renamed to reduce the chance of a naming collision.
+export interface AttributePartConstructorAliases {
+  AttributePart?: ts.Identifier;
+  PropertyPart?: ts.Identifier;
+  BooleanAttributePart?: ts.Identifier;
+  EventPart?: ts.Identifier;
+}
+
+/**
+ * Creates the statement:
+ *
+ * ```ts
+ * import * as <namespace> from "<moduleSpecifier>"
+ * ```
+ */
+const importNamespaceDeclaration = (
+  factory: ts.NodeFactory,
+  namespace: ts.Identifier,
+  moduleSpecifier: ts.StringLiteral
+) =>
+  factory.createImportDeclaration(
+    undefined,
+    factory.createImportClause(
+      false,
+      undefined,
+      factory.createNamespaceImport(namespace)
+    ),
+    moduleSpecifier,
+    undefined
+  );
+
+/**
+ * Add Parts constructors import. To avoid identifier collisions, each part
+ * constructor is mapped to a unique identifier that must be provided. If no
+ * identifier is provided for a part, then the part will not be emitted.
+ *
+ * This uses a namespace import for g3 compatibility.
+ *
+ * ```ts
+ * import * as litHtmlPrivate_1 from "lit-html/private-ssr-support.js";
+ * const { AttributePart: <AttributePart alias>, PropertyPart: <PropertyPart alias>, BooleanAttributePart: <BooleanAttributePart alias>, EventPart: <EventPart alias> } = litHtmlPrivate_1._$LH;
+ * ```
+ */
+export const addPartConstructorImport = ({
+  factory,
+  sourceFile,
+  securityBrand,
+  partIdentifiers,
+}: {
+  sourceFile: ts.SourceFile;
+  securityBrand: ts.Statement;
+  factory: ts.NodeFactory;
+  partIdentifiers: AttributePartConstructorAliases;
+}): ts.SourceFile => {
+  const uniqueLitHtmlPrivateIdentifier =
+    factory.createUniqueName('litHtmlPrivate');
+  const brandIdx = sourceFile.statements.indexOf(securityBrand);
+  if (brandIdx === -1) {
+    throw new Error(
+      `Internal Error: Could not find security brand declaration.`
+    );
+  }
+  // Add parts import immediately before the security brand.
+  const beforeSecurityBrand = sourceFile.statements.slice(0, brandIdx);
+  const afterSecurtyBrand = sourceFile.statements.slice(brandIdx);
+
+  const partConstructors = [
```

Comment:
This can be hoisted outside of the function, and based on `attributePartConstructors`:

```ts
const partConstructors = Object.values(attributePartConstructors);
```

But also, the name is a bit ambiguous here. First, this seems to have only the attribute-part-like constructor names. If the name included "attribute" it'd be very similar to `attributePartConstructors` so my question would be "how is it different?". Then these aren't really constructors, but constructor names.

I tend to just go for long names if they're descriptive and accurate, rather than try too hard to shorten things. I would call this array `attributePartConstructorNames` and the other object possible `attributePartConstructorNameMap`.

Reply:
Much better – done.